onnx/shape_inference/implementation.h

```diff
@@ -241,6 +241,145 @@ struct InferenceContextImpl : public InferenceContext {
+  std::vector<const TypeProto*> allInputTypes_;
+  std::vector<TypeProto> allOutputTypes_;
+  std::unordered_map<std::string, TensorShapeProto>& generatedShapeData_;
+  const GraphInferenceContext* graphInferenceContext_;
+  std::unordered_map<std::string, const AttributeProto*> attributesByName_;
+  std::unordered_map<std::string, GraphProto*> graphProtoAttributesByName_;
+
+  const AttributeProto* getAttribute(const std::string& name) const override {
+    auto iter = attributesByName_.find(name);
+    if (iter == attributesByName_.end()) {
+      return nullptr;
+    } else {
+      return iter->second;
+    }
+  }
+
+  size_t getNumInputs() const override {
+    return allInputTypes_.size();
+  }
+
+  const TypeProto* getInputType(size_t index) const override {
+    if (index >= allInputTypes_.size()) {
+      ONNX_THROW("Input " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    return allInputTypes_[index];
+  }
+
+  size_t getNumOutputs() const override {
+    return allOutputTypes_.size();
+  }
+
+  TypeProto* getOutputType(size_t index) override {
+    if (index >= allOutputTypes_.size()) {
+      ONNX_THROW("Output " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    return &allOutputTypes_[index];
+  }
+
+  // Convert integer vector into TensorShapeProto
+  template <typename INTEGER>
+  TensorShapeProto vectorToTensorShapeProto(const std::vector<INTEGER>& input_vals) const {
+    TensorShapeProto converted_tsp;
+    for (unsigned int i = 0; i < input_vals.size(); ++i) {
+      converted_tsp.mutable_dim()->Add()->set_dim_value(input_vals[i]);
+    }
+    return std::move(converted_tsp);
+  }
+
+  const TensorShapeProto* getInputData(size_t index) override {
+    if (index >= allInputData_.size()) {
+      ONNX_THROW("Input " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    // Gets it from previous data propagation
+    auto iter = generatedShapeData_.find(inputIndexToNameMap_.at(index));
+    if (iter != generatedShapeData_.end()) {
+      return &iter->second;
+    }
+    // Otherwise, gets it from initializer if it exists
+    const auto* input_data = allInputData_[index];
+    if (input_data != nullptr) {
+      // Only scalar (0D tensor) or 1D tensor can be converted
+      if (input_data->dims_size() == 0 || input_data->dims_size() == 1) {
+        TensorShapeProto tsp;
+
+        // Only supports integer type to form a shape
+        if (input_data->data_type() == TensorProto_DataType_INT64) {
+          tsp = vectorToTensorShapeProto(ParseData<int64_t>(input_data));
+        } else if (input_data->data_type() == TensorProto_DataType_INT32) {
+          tsp = vectorToTensorShapeProto(ParseData<int32_t>(input_data));
+        }
+
+        // Adds this TensorShapeProto from initializer into generatedShapeData
+        // for future use
+        generatedShapeData_.insert({inputIndexToNameMap_.at(index), std::move(tsp)});
+        TensorShapeProto* tsp_ptr = &tsp;
```

Comment:
this wont work... you moved tsp on line 364. you should return &generatedShapeData_[inputIndexToNameMap_.at(index)]